crates/persistence_api/src/cas/content.rs

```diff
@@ -0,0 +1,232 @@
+//! Implements a definition of what AddressableContent is by defining Content,
+//! defining Address, and defining the relationship between them. AddressableContent is a trait,
+//! meaning that it can be implemented for other structs.
+//! A test suite for AddressableContent is also implemented here.
+
+use crate::{
+    cas::storage::ContentAddressableStorage,
+    error::PersistenceError,
+    hash::HashString,
+    json::{default_to_json, default_try_from_json, JsonString},
+};
+use multihash::Hash;
+use std::{
+    convert::TryFrom,
+    fmt::{Debug, Write},
+};
+
+/// an Address for some Content
+/// ideally would be the Content but pragmatically must be Address
+/// consider what would happen if we had multi GB addresses...
+pub type Address = HashString;
+
+impl From<Option<Address>> for JsonString {
+    fn from(maybe_address: Option<Address>) -> Self {
+        default_to_json(maybe_address)
+    }
+}
+
+impl TryFrom<JsonString> for Option<Address> {
+    type Error = PersistenceError;
+    fn try_from(j: JsonString) -> Result<Self, Self::Error> {
+        default_try_from_json(j)
+    }
+}
+
+/// the Content is a JsonString
+/// this is the only way to be confident in persisting all Rust types across all backends
+pub type Content = JsonString;
+
+/// can be stored as serialized content
+/// the content is the address, there is no "location" like a file system or URL
+/// @see https://en.wikipedia.org/wiki/Content-addressable_storage
+pub trait AddressableContent {
+    /// the Address the Content would be available at once stored in a ContentAddressableStorage
+    /// default implementation is provided as hashing Content with sha256
+    /// the default implementation should cover most use-cases
+    /// it is critical that there are no hash collisions across all stored AddressableContent
+    /// it is recommended to implement an "address space" prefix for address algorithms that don't
+    /// offer strong cryptographic guarantees like sha et. al.
+    fn address(&self) -> Address {
+        Address::encode_from_str(&String::from(self.content()), Hash::SHA2256)
+    }
+
+    /// the Content that would be stored in a ContentAddressableStorage
+    /// the default implementation covers anything that implements From<Foo> for JsonString
+    fn content(&self) -> Content;
+
+    /// restore/deserialize the original struct/type from serialized Content
+    /// the default implementation covers anything that implements From<JsonString> for Foo
+    fn try_from_content(content: &Content) -> Result<Self, PersistenceError>
+    where
+        Self: Sized;
```

Comment:
Can the content be a slice for example?

Reply:
Open to improving the content api but in a subsequent PR.